backends/ref/ceed-ref-basis.c

```diff
@@ -246,10 +248,19 @@ static int CeedBasisApply_Ref(CeedBasis basis, CeedInt num_elem, CeedTransposeMo
         }
         CeedCallBackend(CeedTensorContractApply(contract, num_comp, P, num_elem, Q, div, t_mode, add, u, v));
       } break;
-      // LCOV_EXCL_START
       // Evaluate the curl to/from the quadrature points
-      case CEED_EVAL_CURL:
-        return CeedError(ceed, CEED_ERROR_BACKEND, "CEED_EVAL_CURL not supported");
+      case CEED_EVAL_CURL: {
+        CeedInt           cdim = (dim < 3) ? 1 : dim;
```

Comment:
What does c in `c_dim` mean? It would be nice to have something at the interface level that other backends can use like `CeedBasisGetNumQuadratureComponents`
```suggestion
        CeedInt           c_dim = (dim < 3) ? 1 : dim;
```

Reply:
`c_dim` is the dimension of the curl, which for 1D and 3D elements is equal to `c_dim = dim` but for 2D elements it is 1D instead of 2D (normal to the reference element plane).

My approach was based on not wanting to keep adding members to the `struct CeedBasis_private` and instead just use information that was already there in the `CeedFESpace` member. I can add back a `CeedBasisGetQuadratureDimension` which returns `q_dim = (fe_space == CEED_FE_SPACE_H1) ? 1 : dim;` and `CeedBasisGetCurlDimension` which returns `c_dim = (dim < 3) ? 1 : dim;` if that's preferable.

As for the naming of `Q_comp` -> `q_dim`, I thought this was just nicer for consistency with `dim` and `c_dim`. I could just as easily have it by `q_comp` and `c_comp` or `curl_comp`, just let me know what you prefer.

Reply:
This ultimately calculation needs to go into at 6 other backends and its used a few places in the interface, so I think its better to have a single source of truth for the calculation.

I'm not an expert in HDiv or HCurl elements, but I was under the impression that the number of components changed, not the dimension of the quadrature space.

Reply:
See [063bea1](https://github.com/CEED/libCEED/pull/1156/commits/063bea108b882e83a5ae894d5a49e974101cbd30).

Reply:
I prefer to use `Q_comp` and `Curl_comp`. @jedbrown what do you prefer?

Reply:
Note that this is `q_comp` and `curl_comp` in the most recent commit 063ea1 to match cases better, but happy to adjust to whatever is desired.

Reply:
That looks good to me, thanks.